### PR TITLE
make `IOStream` thread-safe

### PR DESCRIPTION
**Pull request by JeffBezanson**
Unfortunately this has a hefty ~5x slowdown for small I/O operations.

### CODE REVIEW

**Comment by JeffBezanson** [comment]
We decided to make this thread-safe by default, then if it turns out to be a problem consider adding an option like `open(filename, locked=false)` to disable the lock. Note the overhead is only significant (2x-4x) in the case of very small I/O operations like bytes or small integers. Any code doing large numbers of small IOs can probably be greatly sped up anyway.

**Comment by JeffBezanson** [comment]
Note one reason this is important is that when julia's stdout is redirected to a file, `stdout` is an `IOStream`.

File: NEWS.md [hunk 1]
```diff
@@ -14,6 +14,9 @@ Language changes
 Multi-threading changes
 -----------------------
 
+* All system-level I/O operations (e.g. files and sockets) are now thread-safe.
+  This does not include subtypes of `IO` that are entirely in-memory, e.g. `IOBuffer`.
```
Comment (vtjnash) [inline NEWS.md]:
```suggestion
  This does not include subtypes of `IO` that are entirely in-memory, such as `IOBuffer`,
  although it specifically does include `BufferStream`.
```

**Review by vtjnash — APPROVED** [review]

**Comment by baggepinnen** [comment]
I believe this causes quite the slowdown in WAV.jl
https://github.com/dancasimiro/WAV.jl/issues/52#issuecomment-527745083
Is the `locked=false` option on the horizon?

**Comment by StefanKarpinski** [comment]
Is it possible to read/write in larger units than individual bytes?

**Comment by JeffBezanson** [comment]
Right; while this was a ~5x slowdown; WAV.jl is 40x slower than other wav readers.

**Pull request merged**